File: Contributors/TheMob/2020/flattenArray/flattenArray.js
// Take an array with possible nested arrays, return a 'flattened' version of that array

// For example, with input `[1, 2, 3, [1, 2, 3]]`, return `[1, 2, 3, 1, 2, 3]`.
// With input `[1, 2, 3, [1, 2, [1, 2]]]`, return `[1, 2, 3, 1, 2, 1, 2]`.

// iterate through outer array, identify each item is array -- typeof or Array.isArray()
// use spread operator to get items if an array is found
// continue until no arrays left
// recursion?

const assert = require('assert')


const flattenArray = (array) => {
    if (!Array.isArray(array)) return array
    
    const flattenRecurse = (arr) => {
      let resultArray = []
      for (const item of arr) {
          if (!Array.isArray(item)) {
              // push it to resultArray
              resultArray.push(item)
          } else {
              // item is an array (could be multiple levels)
              resultArray = resultArray.concat(flattenRecurse(item))
          }
      }
      return resultArray
  }
  return flattenRecurse(array)
}

assert.deepStrictEqual(flattenArray([1, 2, 3, [1, 2, 3]]),[1, 2, 3, 1, 2, 3])
assert.deepStrictEqual(flattenArray([1, 2, 3, [1, 2, [1, 2]]]),[1, 2, 3, 1, 2, 1, 2])
assert.deepStrictEqual(flattenArray([1, [2, [3, [4, 5, [6, [7, [8]]]]]]]), [1, 2, 3, 4, 5, 6, 7, 8])
assert.deepStrictEqual(flattenArray(null), null)
console.log('All tests passed!')

File: Contributors/GabrielConlon/flattenArray/flatten_array.py
#!/usr/bin/env python3

"""
Instructions
Take an array with possible nested arrays, return a 'flattened' version of that array

For example, with input [1, 2, 3, [1, 2, 3]],
return [1, 2, 3, 1, 2, 3].

With input [1, 2, 3, [1, 2, [1, 2]]],
return [1, 2, 3, 1, 2, 1, 2].
"""

def flattenArray(arr, flatArray):
    # recursively check each item in the array for type
    # if type == [], flatten
    for idx, item in enumerate(arr):
        if type(item) == list:
            flattenArray(item, flatArray)
        else:
            flatArray.append(item)
    return flatArray

def main():
    test_array = [1, 2, 3, [1, 2, [1, 2]]]
    flatArray = []
    print(test_array, flattenArray(test_array, flatArray))

if __name__ == "__main__":
    main()